walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,104 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.Collections;
+import java.util.Arrays;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Set;
+import java.lang.Math;
 
 public final class FindMeetingQuery {
+
+  private ArrayList<TimeRange> meetings = new ArrayList<TimeRange>();
+  private ArrayList<TimeRange> timeTable = new ArrayList<TimeRange>();
+  private int meetingDuration;
+  private Collection<String> attendees;
+
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+
+    meetingDuration = (int) request.getDuration();
+    attendees = request.getAttendees();
+    
+    // Edge Cases
+    // 1. when requested duration exceeds one day 
+    // 2. there are no attendees requested to the meeting
+    if (request.getDuration() > (60*24)) {
+        return Arrays.asList();
+    } else if (request.getAttendees().size() == 0) {
+        return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+
+    constructTimeTable(events);
+
+    // when there are no timeslots where attendees need to attend events
+    if (timeTable.size() == 0) {
+        return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+    
+    findMeetingSlots(timeTable);
```

Comment:
How do you decide which variables to set as instance variables and which to pass in as arguments to the functions? 

I think instance variables should only be used if there is a need to store state between multiple function calls, which is not the case in this problem. If i'm reading the problem right, there is no need to store state in this problem, i.e. every `.query(...)` call should be independent from each other. 

Functions should be pure functions as much as possible we really need side effects, because pure functions are much easier to reason about. e.g. constructTimetable can return a full timetable instead of modifying the instance variable, and findMeetingSlots can return a list of available slots. Pure functions can prevent subtle bugs that arise from a persistent state, e.g. looking at constructTimetable function, it looks like if `.query` is called twice, the `timetable` will be mixed up since you only add to the timetable variable and never reset it. Same for `meetings`.

Reply:
Hmmm, since in this scenario, for the function constructTimeTable, the variable timeTable is modified in place, I won't need to return it right?